Cleanup language around c_nonce (#525)
openid/OpenID4VCI: openid-4-verifiable-credential-issuance-1_0.md

The Credential Issuer provides a nonce value in the HTTP response with a 2xx status code and the following parameters included as top-level members in the message body of the HTTP response using the application/json media type:

> * `c_nonce`: REQUIRED. String containing a challenge to be used when creating a proof of possession of the key proof (see (#credential-request)). This value MUST be unpredictable, though the Credential Issue MAY use the same value for a time period of it's choosing.

Reviewer: @tplooker are you saying that the Credential Issuer can decide to return the nonce returned in a previous response and at the same time the nonce is unpredictable?

> * `c_nonce`: REQUIRED. String containing a challenge to be used when creating a proof of possession of the key proof (see (#credential-request)). This value MUST be unpredictable, though the Credential Issue MAY use the same value for a time period of it's choosing.

Author: > @tplooker are you saying that the Credential Issuer can decide to return the nonce returned in a previous response and at the same time the nonce is unpredictable?

@andprian Yes. When you pick a new value, the new value selected must not be predictable. How long you use that new value for is (currently) a matter of policy.

Is the expectation now clear to you? Do you have any suggestions to improve the wording?

This is the current situation and this PR only aims to more clearly document the current situation without normative changes. If we need normative changes I think those will most likely come about as a result of further discussion of #461, though if you see other issues please do open an issue for them.

> * `c_nonce`: REQUIRED. String containing a challenge to be used when creating a proof of possession of the key proof (see (#credential-request)). This value MUST be unpredictable, though the Credential Issue MAY use the same value for a time period of it's choosing.

Author: Another suggestion building slightly on Tobias's:

```suggestion
* `c_nonce`: REQUIRED. String containing a challenge to be used when creating a proof of possession of the key proof (see (#credential-request)). It is at the discretion of the Credential Issuer when to return a new challenge value as oppose to the one returned in the previous request. New challenge values MUST be unpredictable.
```

> * `c_nonce`: REQUIRED. String containing a challenge to be used when creating a proof of possession of the key proof (see (#credential-request)). This value MUST be unpredictable, though the Credential Issue MAY use the same value for a time period of it's choosing.

Reviewer 2: I would agree here, unpredictable means that nobody should be able to generate c_nonce for the future, therefore it guarantees freshness. If the c_nonce is also used to create strong replay attack prevention is up to the issuer. To be honest, the stakes are not too high for proof in the Credential Request imo

> * `c_nonce`: REQUIRED. String containing a challenge to be used when creating a proof of possession of the key proof (see (#credential-request)). This value MUST be unpredictable, though the Credential Issue MAY use the same value for a time period of it's choosing.

Reviewer: > > are you saying that the Credential Issuer can decide to return the nonce returned in a previous response and at the same time the nonce is unpredictable?
> 
> @andprian Yes. When you pick a new value, the new value selected must not be predictable. How long you use that new value for is (currently) a matter of issuer policy.
> 
> Is the expectation now clear to you? Do you have any suggestions to improve the wording?
> 
> This is the current situation and this PR only aims to more clearly document the current situation without normative changes. If we need normative changes I think those will most likely come about as a result of further discussion of #461, though if you see other issues please do open an issue for them.

@jogu , I was not requesting normative changes, I was requesting to remove the contradiction. This should be clear not just for us, but for all the people that will read this spec, whether familiar with it or not.
I am ok with the text you suggested, because now it states _in which cases_ the value must be unpredictable.

The proof type contained in the `proofs` parameter is an extension point that enables the use of different types of proofs for different cryptographic schemes.

> The proof(s) in the `proofs` parameter MUST incorporate the Credential Issuer Identifier (audience) and, if the Credential has a `nonce_endpoint`, a `c_nonce` value to allow the Credential Issuer to detect freshness. The way that data is incorporated depends on the key proof type. In a JWT, for example, the `c_nonce` value is conveyed in the `nonce` claim, whereas the audience is conveyed in the `aud` claim. In a Linked Data proof, for example, the `c_nonce` is included as the `challenge` element in the key proof object and the Credential Issuer (the intended audience) is included as the `domain` element.

Author: for consistency:

```suggestion
The proof(s) in the `proofs` parameter MUST incorporate the Credential Issuer Identifier (audience) and, if the Credential has a Nonce Endpoint, a `c_nonce` value to allow the Credential Issuer to detect freshness. The way that data is incorporated depends on the key proof type. In a JWT, for example, the `c_nonce` value is conveyed in the `nonce` claim, whereas the audience is conveyed in the `aud` claim. In a Linked Data proof, for example, the `c_nonce` is included as the `challenge` element in the key proof object and the Credential Issuer (the intended audience) is included as the `domain` element.
```

> The proof(s) in the `proofs` parameter MUST incorporate the Credential Issuer Identifier (audience), and optionally a `c_nonce` value generated by the Credential Issuer to allow the Credential Issuer to detect freshness. The way that data is incorporated depends on the key proof type. In a JWT, for example, the `c_nonce` value is conveyed in the `nonce` claim, whereas the audience is conveyed in the `aud` claim. In a Linked Data proof, for example, the `c_nonce` is included as the `challenge` element in the key proof object and the Credential Issuer (the intended audience) is included as the `domain` element.

Reviewer: Having a sentence saying "MUST ... optionally" is a bit strange. I suggest to split in 2 phrases, one with the mandatory audience and one with an optional nonce.

Author: Removed the optionally:

```suggestion
The proof(s) in the `proofs` parameter MUST incorporate the Credential Issuer Identifier (audience) and, if the Credential has a `nonce_endpoint`, a `c_nonce` value to allow the Credential Issuer to detect freshness. The way that data is incorporated depends on the key proof type. In a JWT, for example, the `c_nonce` value is conveyed in the `nonce` claim, whereas the audience is conveyed in the `aud` claim. In a Linked Data proof, for example, the `c_nonce` is included as the `challenge` element in the key proof object and the Credential Issuer (the intended audience) is included as the `domain` element.
```

> The proof(s) in the `proofs` parameter MUST incorporate the Credential Issuer Identifier (audience), and optionally a `c_nonce` value generated by the Credential Issuer to allow the Credential Issuer to detect freshness. The way that data is incorporated depends on the key proof type. In a JWT, for example, the `c_nonce` value is conveyed in the `nonce` claim, whereas the audience is conveyed in the `aud` claim. In a Linked Data proof, for example, the `c_nonce` is included as the `challenge` element in the key proof object and the Credential Issuer (the intended audience) is included as the `domain` element.

Reviewer 3: ```suggestion
```

I suggest we just remove this paragraph entirely as each concrete proof type already defines this explicitly and this prevents a situation in future where a new proof type is defined that perhaps couldn't satisfy these constraints (albeit unlikely).

A key attestation in JWT format as defined in (#keyattestation-jwt).

> If the Credential Issuer has a Nonce Endpoint (as defined in (#nonce-endpoint)), the `c_nonce` value provided by the Credential Issuer MUST be provided in the key attestation's `nonce` parameter.

Reviewer 2: This is a normative change. Should we mandate nonce endpoint in this case?

Reviewer 3: I agree its better to be consistent here unless there is compelling reason not to be.

Reviewer 2: We don't have any freshness guarantee with this change. Yet again, it's also optional for jwt proof type. So I think I'm okay with this. 

cc @c2bo

* `credential_endpoint`: REQUIRED. URL of the Credential Issuer's Credential Endpoint, as defined in (#credential-request). This URL MUST use the `https` scheme and MAY contain port, path, and query parameter components.

> * `nonce_endpoint`: OPTIONAL. URL of the Credential Issuer's Nonce Endpoint, as defined in (#nonce-endpoint). This URL MUST use the `https` scheme and MAY contain port, path, and query parameter components. If omitted, the Credential Issuer does not require the use of `c_nonce`.

Reviewer: Maybe state that if this is present, the wallet must use this endpoint. I see indirect requirements in various places, for the wallet to use the same value as given in the nonce endpoint but I think we should clearly state that this is a mandatory feature for the wallet, when the Issuer supports it.

## Proof replay {#key-proof-replay}

> If an adversary obtains a key proof (as outlined in #proof-types), they could potentially have a Credential issued that is linked to a key pair controlled by the victim. The `c_nonce` parameter serves as the main defense against pre-generation of key proofs, whilst also limiting the time window for replay attacks. It is RECOMMENDED that Credential Issuers utilize the Nonce Endpoint as specified in (#nonce-endpoint). A Wallet can continue using a given nonce until it is rejected by the Credential Issuer. The Credential Issuer determines for how long a particular nonce can be used. Servers MUST establish a clear policy on whether the same key proof can be reused and for how long, or if each Credential Request requires a new key proof.

Reviewer 3: ```suggestion
If an adversary obtains a key proof (as outlined in #proof-types), they could potentially replay such proof to a Credential Issuer to obtain a duplicate Credential or multiple duplicates issued to the same key. The `c_nonce` parameter serves as the main defense against this, by providing Credential Issuers with a mechanism to determine freshness of a generated proof and thus reject a proof when required. It is RECOMMENDED that Credential Issuers utilize the Nonce Endpoint as specified in (#nonce-endpoint) to protect against this. A Wallet can continue using a given nonce until it is rejected by the Credential Issuer. The Credential Issuer determines for how long a particular nonce can be used.
```

> If an adversary obtains a key proof (as outlined in #proof-types), they could potentially have a Credential issued that is linked to a key pair controlled by the victim. The `c_nonce` parameter serves as the main defense against pre-generation of key proofs, whilst also limiting the time window for replay attacks. It is RECOMMENDED that Credential Issuers utilize the Nonce Endpoint as specified in (#nonce-endpoint). A Wallet can continue using a given nonce until it is rejected by the Credential Issuer. The Credential Issuer determines for how long a particular nonce can be used. Servers MUST establish a clear policy on whether the same key proof can be reused and for how long, or if each Credential Request requires a new key proof.

Reviewer 3: Few suggestions here, the risk isn't key substitution AFAIK, its replaying the key proof and getting multiple duplicates back. Also removed the last sentence as generally the re-use risk is about the nonce/challenge, not the key proof and that is already covered earlier in the paragraph.
